compiler/haskell-ide-core/src/Development/IDE/State/FileStore.hs

```diff
@@ -6,44 +6,74 @@
 module Development.IDE.State.FileStore(
     getFileExists, getFileContents,
     setBufferModified,
-    fileStoreRules
+    fileStoreRules,
+    VFSHandle(..),
+    makeVFSHandle,
+    makeLSPVFSHandle,
     ) where
 
 
 
 import           StringBuffer
 import Development.IDE.Orphans()
 
+import Control.Concurrent.STM
 import qualified Data.Map.Strict as Map
+import Data.Maybe
 import qualified Data.Text as T
 import           Data.Time.Clock
 import           Control.Monad.Extra
 import qualified System.Directory as Dir
 import           Development.Shake
 import           Development.Shake.Classes
 import           Development.IDE.State.Shake
-import           Control.Concurrent.Extra
 import           Control.Exception
 import           GHC.Generics
 import System.IO.Error
 import qualified Data.ByteString.Char8 as BS
 import qualified StringBuffer as SB
 import Development.IDE.Types.Diagnostics
+import qualified Data.Rope.UTF16 as Rope
 import           Data.Time
 
-
--- This module stores the changed files in memory, and answers file system questions
--- from either the memory changes OR the file system itself
-
-type DirtyFiles = Map.Map FilePath (UTCTime, StringBuffer) -- when it was modified, it's current value
-
--- Store the DirtyFiles globally, so we can get at it through setBufferModified
-newtype GlobalDirtyFiles = GlobalDirtyFiles (Var DirtyFiles)
-instance IsIdeGlobal GlobalDirtyFiles
+import Language.Haskell.LSP.Core
+import Language.Haskell.LSP.VFS
+
+-- | haskell-lsp manages the VFS internally and automatically so we cannot use
+-- the builtin VFS without spawning up an LSP server. To be able to test things
+-- like `setBufferModified` we abstract over the VFS implementation.
+data VFSHandle = VFSHandle
+    { getVirtualFile :: Uri -> IO (Maybe VirtualFile)
+    , setVirtualFileContents :: Uri -> T.Text -> Int -> IO ()
```

Comment:
Is the `Int` required? I guess that's the version? Should it not just always increment?

Reply:
In fact I'd just suggest making the `Var` be of type `(Int, Map Uri VirtualFile)` and have a global version.

Reply:
The reason why I kept it this way is that it makes it look a bit more like a reasonable implementation that we could theoretically also use in `damlc ide` if we ever switch to something that does not manage the VFS automagically like `haskell-lsp`. But I guess `haskell-lsp` is unlikely to change in that regard unless so changing it to `Int` seems reasonable.